src/OctoshiftCLI.Tests/HttpDownloadServiceTests.cs

```diff
@@ -14,21 +15,33 @@ public class HttpDownloadServiceTests
     {
         private const string EXPECTED_RESPONSE_CONTENT = "RESPONSE_CONTENT";
         private readonly Mock<OctoLogger> _mockOctoLogger = TestHelpers.CreateMock<OctoLogger>();
+        private readonly Mock<FileSystemProvider> _mockFileSystemProvider = TestHelpers.CreateMock<FileSystemProvider>();
 
         [Fact]
         public async Task Downloads_File()
```

Comment:
I really don't like how I have this test written. It works but it just feels too messy and not specific to the function (too much integration). Definitely open to some ideas here! @ArinGhazarian

Reply:
You can leave the test pretty much as it was before, def. no need to create a a file since this is a unit test and needs to be tested in isolation (no file i/o, no actual network calls, etc.). All you want to do is to make sure that in the end the source stream is getting copied over the target. In other words we want to make sure that the following method call happens with correct paras:
```c#
await streamToReadFrom.CopyToAsync(streamToWriteTo);
```

But to test `CopyToAsync` we need to abstract it out. The easiest way I can think of is to simply add the following method to the `FileSystemProvider`:
```c#
public Task CopySourceToTargetStreamAsync(Stream source, Stream target) => await source.CopyToAsync(target);
```

once you have that, you can simply `Verify` that the calls happen with the correct params:
```c#
_mockFileSystemProvider.Verify(m => m.Open(filePath, FileMode.Create), Times.Once);
_mockFileSystemProvider.Verify(async m => await m.CopySourceToTargetStreamAsync(It.Is<Stream>(x => x.ToArray() == expectedFileContents), It.IsAny<FileStream>(), Times.Once);
```

Also note that you need to mock the `FileSystemProvider.Open` as you did but don't return a specific file stream just return any file stream. Yeah it's not 100% ideal but we're verifying every single call so IMO it's good enough:
```c#
 _mockFileSystemProvider.Setup(x => x.Open(filePath, It.IsAny<FileMode>())).Returns(It.IsAny<FileStream>());
```

Reply:
Addressed in 56688e760716150fa9c4414d0f43c58a1ee74343

The filestream gets closed before the assert statements and I'm not aware of a way around that so I don't think we can verify that the expectedcontents are the same?

Reply:
Well that's a bummer. In that case we can use a [Moq callback](https://github.com/Moq/moq4/wiki/Quickstart#callbacks), so as soon as the params are received we can capture the source stream, convert it to a string and store it in a variable and then in the assertions section additionally assert that the contents are the same as `expectedFileContents`:

```c#
...

string actualFileContents = null;
_mockFileSystemProvider.Setup(x => x.CopySourceToTargetStreamAsync(It.IsAny<Stream>(), It.IsAny<Stream>())).Callback<Stream, Stream>((s, _) =>
{
    using var ms = new MemoryStream();
    s.CopyTo(ms);
    actualFileContents = Encoding.UTF8.GetString(ms.ToArray());
});

// Assert
...

actualFileContents.Should().Be(expectedFileContents);
```

Reply:
Addressed in 7d436e04071bf868990e1d32d6276668c1db1a35 :tada: